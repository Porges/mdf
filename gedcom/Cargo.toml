[package]
name = "gedcom"
version = "0.1.0"
edition = "2021"
authors = ["George Pollard <porges@porg.es>"]
license = "GPL-3.0"

[dependencies]
ascii = "1.1.0"
encoding_rs = "0.8.34"
miette = "7.2.0"
owo-colors = "4.0.0"
thiserror = "1.0.63"
tracing = { version = "0.1", features = ["attributes"] }
vec1 = "1.12.1"

[dev-dependencies]
insta = { version = "1.39.0", features = ["glob"] }
kdl = "4.6.0"
miette = { version = "7.2.0", features = ["fancy-no-backtrace"] }
