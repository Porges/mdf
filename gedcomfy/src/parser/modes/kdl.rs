use kdl::{KdlDocument, KdlEntry, KdlNode};

use crate::parser::{
    lines::LineValue, records::RawRecord, AnySourceCode, NonFatalHandler, ParseMode, ResultBuilder,
    Sourced,
};

#[derive(Default)]
pub(in crate::parser) struct Mode {}

impl NonFatalHandler for Mode {
    fn non_fatal<E>(&mut self, _error: E) -> Result<(), E>
    where
        E: Into<crate::parser::ParseError> + miette::Diagnostic,
    {
        Ok(())
    }
}

impl ParseMode for Mode {
    type ResultBuilder<'i> = Builder;

    fn get_result_builder<'i>(
        self,
        _version: crate::versions::SupportedGEDCOMVersion,
        _source_code: AnySourceCode,
    ) -> Result<Self::ResultBuilder<'i>, crate::parser::ParseError> {
        Ok(Builder {
            mode: self,
            doc: KdlDocument::new(),
        })
    }
}

pub(in crate::parser) struct Builder {
    mode: Mode,
    doc: KdlDocument,
}

impl NonFatalHandler for Builder {
    fn non_fatal<E>(&mut self, error: E) -> Result<(), E>
    where
        E: Into<crate::parser::ParseError> + miette::Diagnostic,
    {
        self.mode.non_fatal(error)
    }
}

impl<'i> ResultBuilder<'i> for Builder {
    type Result = KdlDocument;

    fn handle_record(
        &mut self,
        record: Sourced<RawRecord<'i>>,
    ) -> Result<(), crate::parser::ParseError> {
        self.doc.nodes_mut().push(record_to_kdl(record.value));
        Ok(())
    }

    fn complete(self) -> Result<Self::Result, crate::parser::ParseError> {
        Ok(self.doc)
    }
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.value.to_string()));
    }

    if let Some(mapped) = match record.line.line_value.value {
        LineValue::Ptr(None) => Some(KdlEntry::new_prop("see", kdl::KdlValue::Null)),
        LineValue::Ptr(Some(value)) => Some(KdlEntry::new_prop("see", value)),
        LineValue::Str(data) => Some(KdlEntry::new(data.to_string())),
        LineValue::None => None,
    } {
        node.entries_mut().push(mapped);
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children.nodes_mut().push(record_to_kdl(subrecord.value));
    }

    node.set_children(children);
    node
}
