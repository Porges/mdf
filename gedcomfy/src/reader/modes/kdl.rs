use kdl::{KdlDocument, KdlEntry, KdlNode};

use crate::reader::{
    lines::LineValue, records::RawRecord, NonFatalHandler, ReadMode, ResultBuilder, Sourced,
};

#[derive(Default)]
pub(in crate::reader) struct Mode {}

impl NonFatalHandler for Mode {
    fn report<E>(&mut self, _error: E) -> Result<(), E>
    where
        E: Into<crate::reader::ReaderError> + miette::Diagnostic,
    {
        Ok(())
    }
}

impl<'i> ReadMode<'i> for Mode {
    type ResultBuilder = Builder;

    fn into_result_builder(
        self,
        _version: crate::versions::SupportedGEDCOMVersion,
    ) -> Result<Self::ResultBuilder, crate::reader::ReaderError> {
        Ok(Builder {
            mode: self,
            doc: KdlDocument::new(),
        })
    }
}

pub(in crate::reader) struct Builder {
    mode: Mode,
    doc: KdlDocument,
}

impl NonFatalHandler for Builder {
    fn report<E>(&mut self, error: E) -> Result<(), E>
    where
        E: Into<crate::reader::ReaderError> + miette::Diagnostic,
    {
        self.mode.report(error)
    }
}

impl<'i> ResultBuilder<'i> for Builder {
    type Result = KdlDocument;

    fn handle_record(
        &mut self,
        record: Sourced<RawRecord>,
    ) -> Result<(), crate::reader::ReaderError> {
        self.doc
            .nodes_mut()
            .push(record_to_kdl(record.sourced_value));
        Ok(())
    }

    fn complete(self) -> Result<Self::Result, crate::reader::ReaderError> {
        Ok(self.doc)
    }
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.sourced_value.to_string()));
    }

    if let Some(mapped) = match record.line.value.sourced_value {
        LineValue::Ptr(None) => Some(KdlEntry::new_prop("see", kdl::KdlValue::Null)),
        LineValue::Ptr(Some(value)) => Some(KdlEntry::new_prop("see", value)),
        LineValue::Str(data) => Some(KdlEntry::new(data.to_string())),
        LineValue::None => None,
    } {
        node.entries_mut().push(mapped);
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children
            .nodes_mut()
            .push(record_to_kdl(subrecord.sourced_value));
    }

    node.set_children(children);
    node
}
