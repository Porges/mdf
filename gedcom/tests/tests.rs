use std::{convert::Infallible, path::PathBuf};

use ascii::AsAsciiStr;
use gedcom::{
    Collector, GedcomError, LineStructureError, LineSyntaxError, RawLine, RawRecord, Sink, Sourced,
};
use kdl::{KdlDocument, KdlEntry, KdlNode, KdlValue};
use miette::{Context, NamedSource};

#[test]
fn can_parse_allged_lines() {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/gpdf/allged.ged");

    let data = std::fs::read(path).unwrap();

    let line_count = gedcom::validate_syntax(&data).unwrap();
    assert_eq!(line_count, 1159);
}

#[test]
fn produces_expected_allged_tree() {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/gpdf/allged.ged");

    let data = std::fs::read(path).unwrap();
    let parsed = parse(&data).unwrap();

    insta::assert_snapshot!(to_kdl(parsed.into_iter().map(|r| r.value)));
}

fn parse(data: &[u8]) -> Result<Vec<Sourced<RawRecord<str>>>, miette::Report> {
    let lines = gedcom::iterate_lines(data).map(
        |item| -> Result<(Sourced<usize>, Sourced<RawLine<str>>), LineSyntaxError> {
            // TODO: horrible hack until encoding implemented
            let (l, r) = item?;
            let data = r
                .data
                .as_ref()
                .map(|d| d.map(|s| s.as_ascii_str().unwrap().as_str()));
            let xref = r
                .xref
                .as_ref()
                .map(|d| d.map(|s| s.as_ascii_str().unwrap().as_str()));
            Ok((
                l,
                Sourced {
                    span: r.span,
                    value: RawLine {
                        tag: r.tag,
                        data,
                        xref,
                    },
                },
            ))
        },
    );

    let validated_lines = lines.collect::<Result<Vec<_>, _>>()?;
    let records = gedcom::build_tree(validated_lines.into_iter());
    let validated_records = records.collect::<Result<Vec<_>, _>>()?;

    Ok(validated_records)
}

/*

#[test]
fn torture_test_valid() {
    insta::glob!("tests/external/TestGED/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap().to_string_lossy();
        let result = gedcom::validate(&data).with_context(|| format!("validating {}", filename));
        match result {
            Ok(tree) => insta::assert_debug_snapshot!(tree),
            Err(err) => insta::assert_snapshot!(format!(
                "{:?}",
                err.with_source_code(NamedSource::new(filename, data))
            )),
        };
    });
}

*/
 */

#[test]
fn golden_files() -> miette::Result<()> {
    miette::set_hook(Box::new(|_diag| {
        Box::new(
            miette::MietteHandlerOpts::new()
                .terminal_links(false)
                .unicode(true)
                .color(false)
                .width(132)
                .build(),
        )
    }))?;

    insta::glob!("syntax_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap();
        match parse(&data) {
            Ok(records) => {
                let kdl = to_kdl(records.into_iter().map(|r| r.value));
                insta::assert_snapshot!(kdl);
            }
            Err(err) => insta::assert_snapshot!(format!(
                "{:?}",
                err.with_source_code(NamedSource::new(filename.to_string_lossy(), data))
            )),
        };
    });

    insta::glob!("format_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap().to_string_lossy();
        match parse(&data) {
            Ok(records) => {
                let kdl = to_kdl(records.into_iter().map(|r| r.value));
                insta::assert_snapshot!(kdl);
            }
            Err(err) => insta::assert_snapshot!(format!(
                "{:?}",
                err.with_source_code(NamedSource::new(filename, data))
            )),
        };
    });

    Ok(())
}

fn to_kdl<'a>(records: impl Iterator<Item = RawRecord<'a>>) -> KdlDocument {
    let mut doc = KdlDocument::new();
    for record in records {
        doc.nodes_mut().push(record_to_kdl(record));
    }

    doc
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.value.to_string()));
    }

    if let Some(data) = &record.line.data {
        node.entries_mut()
            .push(KdlEntry::new(data.value.to_string()));
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children.nodes_mut().push(record_to_kdl(subrecord.value));
    }

    node.set_children(children);
    node
}
