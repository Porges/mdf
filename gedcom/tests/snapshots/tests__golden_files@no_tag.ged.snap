---
source: gedcom/tests/tests.rs
expression: "format!(\"{:?}\",\n    err.with_source_code(NamedSource::new(filename.to_string_lossy(), data)))"
input_file: gedcom/tests/syntax_inputs/no_tag.ged
---
gedcom::parse_error::no_tag

  × No tag found
   ╭─[no_tag.ged:1:1]
 1 │ 0
   · ┬
   · ╰── no tag in this line
   ╰────
